### ISSUE TITLE
Violation in times(...) expectations of metastore mocks might not lead to a test failure.

### ISSUE THREAD

**Issue by imotov** [quickwit-oss/quickwit#2754]
**Describe the bug**
When a mock metastore is passed in Arc to an actor, it is not getting finalized until the actor shutdown. The shutdown operation is performed inside tokio task and as a result panic that mock metastore generates if expectations are violated is only printed on screen but doesn't lead to the test failure.

**Steps to reproduce (if applicable)**

1. Replace the following line https://github.com/quickwit-oss/quickwit/blob/eb989e2df2c841dcac275c07e8877b2569b36175/quickwit/quickwit-indexing/src/actors/publisher.rs#L280 with `.times(10)`
2. Run the test
3. Test prints out the [following message](https://gist.github.com/imotov/5df6e837c1a61a6f5cfab97dd9a21770) but passes.

**Expected behavior**
Test should fail.

**Configuration:**
I ran it on 7494137c54cbe08e73ce73556f1b991e166b01c4

**Possible Solution.**
It seems that explicitly quitting the actor that we passed the metastore into and checking its exit status can helps us to identify this situation. Adding the following line at the end of the test, makes the test to fail if mock expectations fail, as expected:
```
assert!(matches!(publisher_handle.quit().await.0, ActorExitStatus::Quit));
```

**Comment by guilload** [comment]
Something like `Universe.join()` could also be useful.

**Comment by imotov** [comment]
So what is the main intent of this method? Is it just join on all spawned actors or is it more like quit, where we shut them down first? And what should I do with the exit statuses? Return them to the caller as a vector? Check for panics and panic?

**Comment by imotov** [comment]
@guilload I have a draft implementation of it. I think the biggest struggle for me is what to do with ActorHandle's quit, kill and join methods. When I call universe.quit() it quits all actors, but since ActorHandles might still exist somebody can call any of these methods on them even though underlying JoinHandler is now gone. At the moment I am preserving and returning ActorExitStatus that was generated during universe.quit(), but I wonder if I should panic instead if somebody tries to call ActorHandle's method after universe.quit();

**Comment by imotov** [comment]
I opened #2766 to demonstrate one of the approaches more clearly.

**Comment by imotov** [comment]
I think this can be closed since #2766 now ensures that at least in case of universe where most of these mocks are used.

**Comment by fulmicoton** [comment]
Thank you @imotov !
